README.md
Mojo::JSON::MaybeXS
=================

Module to use JSON::MaybeXS as the JSON renderer for Mojolicious.


Changes
{{$NEXT}}

1.002     2019-08-07 13:01:24 EDT
  - Set allow_dupkeys when using Cpanel::JSON::XS >= 4.09
  - Doc updates for Mojo::JSON using Cpanel::JSON::XS natively in Mojolicious 7.87+

1.001     2018-02-06 14:06:23 EST
  - Doc updates for upgraded numbers logic

1.000     2017-08-02 00:12:05 EDT
  - Fix tests and document that Cpanel::JSON::XS now disallows duplicate keys in non-relaxed mode

0.012     2016-06-30 17:44:11 EDT
  - Enable canonical option for all backends to match Mojo::JSON behavior in Mojolicious 6.58+

0.011     2016-06-02 14:15:24 EDT
  - escape_slash option will now be enabled if the backend is JSON::PP or Cpanel::JSON::XS >= 3.0206
  - Fix tests to handle trailing .0 added to NV ("numeric non-integer") by Cpanel::JSON::XS >= 3.0214

0.010     2015-12-02 14:54:33 EST
  - Update tests for blessed reference stringification in Cpanel::JSON::XS 3.0207

0.009     2015-09-06 17:32:56 EDT
  - Change installer to Module::Build::Tiny
  - Better examples in synopsis for use with morbo and hypnotoad

0.008     2015-07-13 20:25:16 EDT
  - Update distribution metadata

0.007     2014-12-14 16:49:55-05:00 America/New_York
  - Added Cpanel::JSON::XS 3.0112 to recommended prereqs
  - Enabled new Cpanel::JSON::XS stringify_infnan option where supported

0.006     2014-12-12 10:21:25-05:00 America/New_York
  - Preload JSON booleans for true and false subs
  - Updated tests and documentation for Cpanel::JSON::XS 3.0108
    - inf and nan encoding
    - upgraded number encoding
    - removed doc section on boolean stringification differences

0.005     2014-12-09 18:05:21-05:00 America/New_York
  - Enable allow_unknown option
  - Update docs regarding handling of unblessed references
  - Restore tests related to scalar references

0.004     2014-12-05 12:30:04-05:00 America/New_York
  - Removed Test::Without::Module testing dependency

0.003     2014-11-30 22:53:18-05:00 America/New_York
  - Cleaned up tests and imports

0.002     2014-11-30 18:52:20-05:00 America/New_York
  - Improved testing for JSON::XS and JSON::PP

0.001     2014-11-30 02:27:46-05:00 America/New_York
  - First release


Build.PL
# This Build.PL for Mojo-JSON-MaybeXS was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for Mojo-JSON-MaybeXS was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.010001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.012.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "use JSON::MaybeXS as the JSON encoder for Mojolicious",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "Mojo-JSON-MaybeXS",
      "dist_version" => "1.002",
      "license" => "artistic_2",
      "module_name" => "Mojo::JSON::MaybeXS",
      "recommends" => {
        "Cpanel::JSON::XS" => "3.0207"
      },
      "recursive_test_files" => 1,
      "requires" => {
        "JSON::MaybeXS" => "1.001",
        "Mojolicious" => "5.66",
        "perl" => "5.010001"
      },
      "test_requires" => {
        "File::Spec" => 0,
        "Module::Metadata" => 0,
        "Scalar::Util" => 0,
        "Test::More" => "0.88"
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "Module::Metadata" => 0,
      "Scalar::Util" => 0,
      "Test::More" => "0.88"
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.037.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "use JSON::MaybeXS as the JSON encoder for Mojolicious",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojo-JSON-MaybeXS",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Mojo::JSON::MaybeXS",
  "PREREQ_PM" => {
    "JSON::MaybeXS" => "1.001",
    "Mojolicious" => "5.66"
  },
  "TEST_REQUIRES" => {
    "Scalar::Util" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.008",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "JSON::MaybeXS" => "1.001",
  "Mojolicious" => "5.66",
  "Scalar::Util" => 0,
  "Test::More" => "0.88"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/JSON/MaybeXS.pm
package Mojo::JSON::MaybeXS;

use strict;
use warnings;
use Mojo::Util 'monkey_patch';
use JSON::MaybeXS 'JSON';
use Mojo::JSON ();

our $VERSION = '1.003';

my $BINARY = JSON::MaybeXS->new(utf8 => 1, canonical => 1, allow_nonref => 1,
	allow_unknown => 1, allow_blessed => 1, convert_blessed => 1);
my $TEXT = JSON::MaybeXS->new(utf8 => 0, canonical => 1, allow_nonref => 1,
	allow_unknown => 1, allow_blessed => 1, convert_blessed => 1);
my $TRUE = JSON->true;
my $FALSE = JSON->false;

if (JSON eq 'Cpanel::JSON::XS') {
	local $@;
	if (eval { Cpanel::JSON::XS->VERSION('4.09'); 1 }) {
		$BINARY->allow_dupkeys;
		$TEXT->allow_dupkeys;
	}
	if (eval { Cpanel::JSON::XS->VERSION('3.0112'); 1 }) {
		$BINARY->stringify_infnan;
		$TEXT->stringify_infnan;
	}
	if (eval { Cpanel::JSON::XS->VERSION('3.0206'); 1 }) {
		$BINARY->escape_slash;
		$TEXT->escape_slash;
	}
} elsif (JSON eq 'JSON::PP') {
	$BINARY->escape_slash;
	$TEXT->escape_slash;
}

monkey_patch 'Mojo::JSON', 'encode_json', sub { $BINARY->encode($_[0]) };
monkey_patch 'Mojo::JSON', 'decode_json', sub { $BINARY->decode($_[0]) };

monkey_patch 'Mojo::JSON', 'to_json',   sub { $TEXT->encode($_[0]) };
monkey_patch 'Mojo::JSON', 'from_json', sub { $TEXT->decode($_[0]) };

monkey_patch 'Mojo::JSON', 'true',  sub () { $TRUE };
monkey_patch 'Mojo::JSON', 'false', sub () { $FALSE };

1;

=head1 NAME

Mojo::JSON::MaybeXS - use JSON::MaybeXS as the JSON encoder for Mojolicious

=head1 SYNOPSIS

 use Mojo::JSON::MaybeXS;
 use Mojo::JSON qw/encode_json decode_json true false/;
 
 # Preload for scripts using Mojo::JSON
 $ perl -MMojo::JSON::MaybeXS -S morbo myapp.pl
 
 # Must be set in environment for hypnotoad
 $ PERL5OPT=-MMojo::JSON::MaybeXS hypnotoad myapp.pl

=head1 DESCRIPTION

L<Mojo::JSON::MaybeXS> is a monkey-patch module for using L<JSON::MaybeXS> as
the JSON encoder for a L<Mojolicious> application, or anything else using
L<Mojo::JSON>. It must be loaded before L<Mojo::JSON> so the new functions will
be properly exported.

Since L<Mojolicious> version 7.87, L<Mojo::JSON> has delegated to
L<Cpanel::JSON::XS> by default if installed and recent enough. Installing
L<Mojolicious> version 7.87+ and L<Cpanel::JSON::XS> version 4.09+ resolves the
below listed caveats between these modules, and is sufficient to improve the
performance of L<Mojo::JSON> without the use of this module.

=head1 CAVEATS

L<JSON::MaybeXS> may load different modules behind the scenes depending on what
is available, and these modules have slightly different behavior from
L<Mojo::JSON> and occasionally from each other. References to the behavior of
L<JSON::MaybeXS> below are actually describing the behavior shared among the
modules it loads.

L<JSON::MaybeXS> is used with the options C<canonical>, C<allow_nonref>,
C<allow_unknown>, C<allow_blessed>, and C<convert_blessed>. C<canonical>
enables sorting of hash keys when encoding to JSON objects as L<Mojo::JSON>
does. C<allow_nonref> allows encoding and decoding of bare values outside of
hash/array references, since L<Mojo::JSON> does not prevent this, in accordance
with L<RFC 7159|http://tools.ietf.org/html/rfc7159>. The other options prevent
the encoder from blowing up when encountering values that cannot be represented
in JSON to better match the behavior of L<Mojo::JSON>. See below for more
specifics.

To better match the behavior of L<Mojo::JSON>, certain options may be enabled
depending on the backend that is used. If L<Cpanel::JSON::XS> version 3.0112 or
greater is loaded, it will be used with the option C<stringify_infnan>. If
either L<Cpanel::JSON::XS> of at least version 3.0206 or L<JSON::PP> is loaded,
it will be used with the option C<escape_slash>. If L<Cpanel::JSON::XS> version
4.09 or greater is loaded, it will be used with the option C<allow_dupkeys>.

As of this writing, the author has found the following incompatibilities:

=head2 Object Conversion

Both L<JSON::MaybeXS> and L<Mojo::JSON> will attempt to call the TO_JSON method
of a blessed reference to produce a JSON-friendly structure. If that method
does not exist, L<Mojo::JSON> or L<Cpanel::JSON::XS> version 3.0207 or greater
will stringify the object, while L<JSON::XS> or L<JSON::PP> will always encode
it to C<null>.

 print encode_json([DateTime->now]);
 # Mojo::JSON or Cpanel::JSON::XS >= 3.0207: ["2014-11-30T04:31:13"]
 # JSON::XS or JSON::PP: [null]

=head2 Unblessed References

L<JSON::MaybeXS> does not allow unblessed references other than to hashes,
arrays, or the scalar values C<0> and C<1>, and will encode them to C<null>.
Before L<Mojolicious> version 7.87, L<Mojo::JSON> will treat all scalar
references the same as references to C<0> or C<1> and will encode them to
C<true> or C<false> depending on their boolean value, and other references
(code, filehandle, etc) will be stringified.

Since L<Mojolicious> version 7.87, L<Mojo::JSON>'s behavior with unblessed
references is the same as L<JSON::MaybeXS>.

 print encode_json([\'asdf', sub { 1 }]);
 # Mojo::JSON (Mojolicious >= 7.87): [null,null]
 # JSON::MaybeXS: [null,null]

=head2 Escapes

L<Mojo::JSON> currently escapes the slash character C</> for security reasons.
Before L<Mojolicious> version 7.87, it also escaped the unicode characters
C<u2028> and C<u2029>. L<Cpanel::JSON::XS> version 3.0206 or greater and
L<JSON::PP> will have the option set to escape the slash character, and
L<JSON::XS> does not escape these characters. This does not affect decoding of
the resulting JSON.

 print encode_json(["/\x{2028}/\x{2029}"]);
 # Mojo::JSON (Mojolicious >= 7.87): ["\/ \/ "]
 # Cpanel::JSON::XS >= 3.0206 or JSON::PP: ["\/ \/ "]
 # JSON::XS: ["/ / "]
 # Both decode to arrayref containing: "/\x{2028}/\x{2029}"

=head2 inf and nan

L<Mojo::JSON> encodes C<inf> and C<nan> to strings. L<Cpanel::JSON::XS> version
3.0112 or greater will also stringify C<inf> and C<nan>. However, L<JSON::XS>
or L<JSON::PP> will encode them as numbers (barewords) producing invalid JSON.

 print encode_json([9**9**9, -sin 9**9**9]);
 # Mojo::JSON or Cpanel::JSON::XS >= 3.0112: ["inf","nan"] (on Linux)
 # JSON::XS or JSON::PP: [inf,nan]

=head2 Upgraded Numbers

L<JSON::MaybeXS>, if using L<JSON::XS>, will attempt to guess if a value to be
encoded is numeric or string based on whether Perl has ever populated a string
value for it internally. Therefore, using a variable containing C<13> in a
string context will cause it to be encoded as C<"13"> even if the variable
itself was not changed. L<Mojo::JSON>, L<JSON::PP> version 2.92 or greater, or
L<Cpanel::JSON::XS> version 3.0109 or greater will encode C<13> as C<13>
regardless of whether it has been used as a string.

 my ($num1, $num2) = (13, 14);
 my $str = "$num1";
 print encode_json([$num1, $num2, $str]);
 # Mojo::JSON, JSON::PP >= 2.92, Cpanel::JSON::XS >= 3.0109: [13,14,"13"]
 # JSON::XS: ["13",14,"13"]

=head2 Duplicate Keys

L<Mojo::JSON>, L<JSON::XS>, and L<JSON::PP> will silently accept duplicate keys
in the same JSON object when decoding a JSON string. L<Cpanel::JSON::XS>
version 3.0235 or greater will throw an exception if duplicate keys are
encountered. L<Cpanel::JSON::XS> version 4.09 or greater will have the option
set to once again accept duplicate keys.

 print dumper decode_json('{"foo":1, "bar":2, "foo":3}');
 # Mojo::JSON, JSON::XS, or JSON::PP: { bar => 2, foo => 3 }
 # Cpanel::JSON::XS >= 3.0235 and < 4.09: "Duplicate keys not allowed" exception

=head1 BUGS

This is a monkey-patch of one of a few possible modules into another, and they
have incompatibilities, so there will probably be bugs. Report any issues on
the public bugtracker.

=head1 AUTHOR

Dan Book, C<dbook@cpan.org>

=head1 CREDITS

Sebastian Riedel, author of L<Mojolicious>, for basic implementation.

=head1 COPYRIGHT AND LICENSE

Copyright 2014, Dan Book.

This library is free software; you may redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Mojo::JSON>, L<JSON::MaybeXS>, L<Cpanel::JSON::XS>, L<JSON::XS>, L<JSON::PP>
